[RDY] tests: Migrate legacy test 62.

lucc opened the pull request:
There is still one TODO note in the test file: At some point the "tab drop" command is tested _but_ only `if has('gui') || has('clientserver')`.  Neovim has neither but the test lines that are guarded like this seem to work.  Now I appended `|| has('nvim')` to that check.

The main problem at this point that the `tab drop some-file` command syntax is not documented in the help files (in vanilla ViM `helpgrep tab drop` did not turn up anything useful besides a note in version7.txt).  Does anyone know if there is some documentation on this?

lucc committed:
tests: Migrate legacy test 62.

justinmk commented on test/functional/legacy/062_tab_pages_spec.lua:
```diff
@@ -0,0 +1,330 @@
+    -- TODO: which of these conditions in the first if are necessary?
+    -- Currently we override them all with " || has('nvim')".
+    source([[
+      if has('gui') || has('clientserver') || has('nvim')
```
You can remove the `if` check entirely.

splinterofchaos commented on test/functional/legacy/062_tab_pages_spec.lua:
```diff
@@ -0,0 +1,330 @@
+      q
+      call append(line('$'), 'tab page ' . nr)
+      unlet nr
+    ]])
```
We can simplify this test by removing  "tab page 2" from the results and instead putting
```lua
execute('tabnew')
eq(2, eval('tabpagenr()'))
execute('q')
```
here, and similar for the other sections.

splinterofchaos replied:
Some of these are simple enough that I'd appreciate this change, but I don't want to burden you by enforcing it. Please use your best judgement.

splinterofchaos commented on test/functional/legacy/062_tab_pages_spec.lua:
```diff
@@ -0,0 +1,330 @@
+-------
+
+    execute('for i in range(9) | tabnew | endfor')
+    feed('1gt<cr>')
```
`1gt` is a normal mode command and does not require the `<cr>`,

lucc replied:
You where right in this case (in general enter in normal mode command moves to the next line).

lucc committed:
tests: Modernize migrated legacy test 62.

- use eval() and eq() in many places instead of writing to the buffer
- remove has('autocmd') checks and use corresponding code unconditionally as
  neovim always has the autocmd feature
- split the test into several it() blocks

Helped-By: Scott Prager <splinterofchaos@gmail.com>
Helped-By: Michael Reed <m.reed@mykolab.com>

lucc commented:
Update: I removed the TODO (and the if line).

@splinterofchaos where did you get the `eq()` function? I am currently using `assert.equals()`. I like the idea very much. I hope nobody comes up and says that it is not a good idea for the legacy tests.

splinterofchaos commented:
```lua
local eq, neq = helpers.eq, helpers.neq
```
> I hope nobody comes up and says that it is not a good idea for the legacy tests.

I believe a large part of the motivation is to make them simpler and more readable. To me, that means avoiding this `redir` stuff and making them less stateful.

lucc commented:
@splinterofchaos I found the eq function and friends. Thanks. I also tried to use it in as many places as it seemed worth it.

splinterofchaos commented:
Sorry, I almost forgot to come back and review this! LGTM :+1:

Don't know if you plan to squash the two commits, but I think we should keep them separate. The first translates the test somewhere literally, and the second refractors it.

lucc commented:
rebased and minor style improvements.

I will now mark this as RDY.

lucc commented:
ref for a better overview: #1286

splinterofchaos commented:
Sorry for the long delay on this, I wanted to give other people a chance to review it. Any takers?

Pyrohh commented on test/functional/legacy/062_tab_pages_spec.lua:
```diff
@@ -0,0 +1,272 @@
+      let g:r=[]
+      command -nargs=1 -bar C :call add(g:r, '=== '.<q-args>.' ===')|<args>
+      function Test()
+          let hasau=has('autocmd')
```
This should always evaluate to `1`, so should be removed. The conditionals below should be adjusted as well.

Pyrohh commented on test/functional/legacy/062_tab_pages_spec.lua:
```diff
@@ -0,0 +1,272 @@
+  it('is working', function()
+    insert('Results:')
+
+    execute('lang C')
```
@lucc @splinterofchaos Do either of you know if this is needed?

lucc replied:
It doesn't fail for me but from `:h :lang` I infer that it is needed for the test of `E474` in this test file. Maybe someone with a non English system or more knowledge can say more.

Pyrohh commented on test/functional/legacy/062_tab_pages_spec.lua:
```diff
@@ -0,0 +1,272 @@
+    eq(2, eval('tabpagewinnr(2, "$")'))
+    eq(1, eval('tabpagewinnr(2)'))
+    execute('tabclose')
+    execute('q')
```
Could these tests be split up into `it()` blocks? I think it would avoid the need for `execute('q')` and/or `execute('tabclose')`, but I'm not too experienced with these tests so idk.

lucc commented:
@Pyrohh I incorporated your comments. I will squash the last three commits into one commit later on: "tests: Modernize migrated test 62." Just tell me if you think any of the three commits is not worth it.

Pyrohh commented:
The changes look good to me, although I wonder why the tests are failing. @lucc I can't run the tests locally at the moment, but are thy passing on your machine?

splinterofchaos commented:
I looked at the failures, but they don't seem related at all since this PR only modifies one test and others are failling.

lucc commented:
I also deem the travis failures unrelated. I get 450 passing and one pending test on master and 458 passing and one pending test on test62 if I run them locally.

There have been similar unrelated test failures in the past: #2841

Pyrohh commented:
Well in that case, LGTM.

lucc commented:
The test still fail in an unrelated test file but in a different one after you restarted the tests: First it was a failure in [`test/functional/ui/searchhl_spec.lua`](https://travis-ci.org/neovim/neovim/jobs/70126324) and [`test/functional/clipboard/clipboard_provider_spec.lua` and `test/functional/ui/syntax_conceal_spec.lua`](https://travis-ci.org/neovim/neovim/jobs/70126331). Now it is [`test/functional/legacy/107_adjust_window_and_contents_spec.lua`](https://travis-ci.org/neovim/neovim/jobs/70180890) ([twice](https://travis-ci.org/neovim/neovim/jobs/70180891)).

Pyrohh commented:
@lucc I think you're right; that `tty ready` issue has been around for a while, but just to be sure:
@splinterofchaos @justinmk do you have any objections before merging?

splinterofchaos commented:
Lgtm :+1;

Pyrohh commented:
Thanks @splinterofchaos.

@lucc Please rebase and squash your commits into 2, as well as mentioning the changes made in the 2nd commit (the modernization one). It might not be obvious why the `has('autocmd')` changes were made, for example.

justinmk commented:
`gcda` test failures (ctrl-f "gcda" in the build log) are a gcov quirk. It's nothing to do with your PR.

justinmk commented on test/functional/legacy/062_tab_pages_spec.lua:
```diff
@@ -0,0 +1,240 @@
+describe('tab pages', function()
+  before_each(clear)
+
+  it('can be opend and closed', function()
```
typo

lucc replied:
done

lucc commented:
squased and rebased

Pyrohh commented:
Thanks @lucc.
